### ISSUE TITLE
Double screen and fullscreen mode: issue with ESC and CMD+P on MACOSX+GAMA Git and CI

### ISSUE THREAD

**Issue by benoitgaudou** [gama-platform/gama#2479]
Hi,

**Configuration:**
- MAC OSX, GAMA Git up-to-date
- Mac BookPro laptop
- TV Screen pluged with HDMI, with extended desktop
- the preference ""Disable the OS menu bar when a display is full-screen"" is on off.

**The model:** 
```
model multiscreen

global {
 init{
 	create people number:100;
 }
}

species people{
	
	reflex t {location <- any_location_in(world);}
	aspect base{
		draw circle(1);
	}
}
experiment multiscreen type: gui{
	output {		
		
        display CityScope type:opengl{
        	species people aspect: base;
        }
		display CityScopeTable   type:opengl fullscreen:1 {
			species people aspect: base;
		}
	}
}
```

**Use case:**
- I run the model from GAMA on the "main" screen.
The two screens appear.
On the full screen: 
  - the CMD+P, the ESC do NOT work.
  - But all the mouse actions work: left-click to move the camera, right-click to open contextual menu, all the Presentation menu (side control, overlay and toggle bar) work.

**What can make the CMD+P and ESC works**
- When I click on another application, Eclipse, Chrome ... , and click back on GAMA, now CMD+P and the eSC work.
- When I open, in the fullscreen, the toolbar, and use the bottom-right button to browse all the agents of the species people.
- When I launch first a Srirama's "magic" model... 
when I launch first the following model, close it and then launch the Fullscreen above model, the CMD+P works. Notice that this does not work with all the models ...
```
model skeleton

global {
	list<point> my_points;
	geometry my_convex_hull;
	geometry my_polygon;

	init {
		my_points <- [{10, 10}, {13, 43}, {97, 32}, {54, 76}];//, {10, 10}];
			
		my_polygon <- polygon(my_points);
		write my_polygon;
		write skeletonize(my_polygon);
		
		my_convex_hull <- convex_hull(my_polygon);
		write my_convex_hull;
		write skeletonize(my_convex_hull);
	}
}

experiment name type: gui {
	output {
		display d {
			graphics "g" {
				draw my_polygon empty: true color: #red;
				draw envelope(my_polygon) /*my_convex_hull*/ empty: true color: #green;
			} } } }
```

**What do not change anything conerning the CMD+P and ESC**
- When I browse all the agents, that makes the CMD+P works. BUT when I inspect a single agent (by clicking on the agent in the fullscreen display or from the bottom-right button), this does not change anything with the CMD+P.
- Running first Srirama's model, and then the fullscreen model, CMD+P works. But with the first model of the model library (3D camera and trajectories.gaml) first and after the Fullscreen model, this does not work.

When I toggle the preference ""Disable the OS menu bar when a display is full-screen""  on on.
  I have exactly the same bahavior.

Benoit

**Comment by sriramab** [comment]
Dont want to mess up this issue, my only point is, click other application and come back to GAMA does not solve my ESC problem. I still have to force quit as  I donot get simulation controls on the toolbar.

**Comment by benoitgaudou** [comment]
I updated my issue to keep a clear picture of what works and not.
I tested : 
- with GAMA Continuous Built (no change compared to the Git)
- I found new ways to make the CMD+P work

**Comment by AlexisDrogoul** [comment]
OK. Bad news. Because mine is working out of the box without problem. Almost the same setup, same preferences, same version of GAMA. CMD-P, ESC, all the mouse controls, toolbar at the bottom, everything works **except** something that can effectively be a bit surprising with 2 screens: the "regular" GAMA does not react to anything unless the fullscreen display is disabled. But it was the intended behaviour from the beginning (i.e. it has always been like that AFAIK), to prevent users from messing up with the fullscreen configuration. A short movie to show the thing in action:

https://photos.app.goo.gl/8YuhroiguRx4KghBA

So, despite my good will (I installed a secondary screen just for this bug), I don't see what I can do. The only differences I see are : 

 - I'm not using the CI build but the Eclipse version
- My connection to the screen is VGA, not HDMI. 

I will try with the CI build and report here.

**Comment by hqnghi88** [comment]
I propose to whom has blocking state, could you please retry the same model but comment the fullscreen (or make it false), to see if there is an exception or any strange behavior?

**Comment by hqnghi88** [comment]
After 1 hour experience on Sriri's machine, I have something to say incrementally:
BIG NOTE: on macos, right click on an app in bottom bar, when we have several display, it can be set in pop menu option there. I chose All Display. it explain why Sriri screenshot is a floating windows, he chose display 1, so fullscreen on second screen is nowhere, just on display 0 (main).
- I first try with his eclipse, git. Launch his model (multiscreen above). Esc work.
- As he said his CI did not work, I tried, and blocked, with many combination preferences, code...
- The same CI, i create new workspace, and copy the model, voila. ESC work.
- I return to old workspace, comment the fullscreen facet and launch the same model. and you guess what, i get #2441 an abnormal exception. Then i try some preferences about layout (even None, Stack) but none help. 
-Then For the same model, i change the name of the display (CityScope0, and CityScopeTable1), i can launch again without "abnormal". Then fullscreen mode work again (esc can esc).
- With CityScope0, and CityScopeTable1, i retry preference Layout Horizontal. after some launch, it recreate "abnormal.....". Then i change display name again (CityScope00, and CityScopeTable01), again it can escape.

So , i think the focus is randomly catch somewhere in the exception message view.

**Comment by AlexisDrogoul** [comment]
@hqnghi88 : thanks for this remote debugging. However, and this is something we should maybe clarify (i.e. forbid), layouts **are not supposed** to work with fullscreen displays. We can go fullscreen when a layout has been applied, but defining both a layout **and** a fullscreen display is sure to create problems. 
My guess (but it's just a guess) is that issue #2441 happening in the old workspace without fullscreen is **because** of a previous blocking (and the necessity to force-quit), which saves a dirty perspective in the metadata, which in turn makes garbage when trying to recreate views with the same id (hence the changing of names which help). 
So the two bugs seem to interact: one one hand, force-quitting when fullscreen blocks creates the condition for #2441, which, after that, probably creates the condition for this issue to manifest itself when in fullscreen (bad initialisation of the viewpart, I don't know). 
What is surprising however in your report is that the Git version was working while the CI was not on the same workspace... ?

**Comment by hqnghi88** [comment]
I want to clarify that i am not supposed to use layout as in my 2nd step, i tried with a non-working already. <b> So the git should have the same if i kept trying (may be). </b> 
Later, i have try to change several times the display names . The consequence when "abnormal" is the blocked fullscreen. and of course there is NO ``fullscreen force quit create the "abnormal" `` ( or you mean another cause of force quit?) as i just relaunch several time without fullscreen and it create "abnormal"

**Comment by hqnghi88** [comment]
And i wait these two confirm the non-fullscreen state, as they tried many models and i think if the focus is in a exception view (blocking cause), it should be randomly in any model they have meshed with. They did not mention new workspace (but even new workspace, if they play with layout, it should have occured).

**Comment by AlexisDrogoul** [comment]
@hqnghi88 I am currently trying out a strategy for #2241. We can continue the discussion in this issue.

@benoitgaudou @agrignard I will issue soon a tentative commit for this issue. I am still fighting with a small problem with the toolbar, but it shouldn't take too much time.

**Comment by hqnghi88** [comment]
You are right Alexis, the git should not supposed to work, i just remembered that i tried git on the same model of the same workspace (which give immediatly issue in CI). The only reason i think is the workspace of configuration are different, it was the place to save the cache of application, for git version is the workspace of eclipse (a clean, not occur yet), and the gama application (CI) have already a dirty on.

**Comment by AlexisDrogoul** [comment]
@hqnghi88 you are perfectly right. The space for the metadata is different. 

I have committed what I hope will be a fix for this. Note that the behaviour of the toolbar is subtly different (but much more coherent with the non fullscreen state): to obtain a fullscreen w/o toolbar, one has to write now:

```
display xxx type: yyy fullscreen: 1 toolbar: false {...}
```

Or to turn the toolbar off globally in the preferences. 

Please test when you have time and report, I hope, successes... 🤔

**Comment by benoitgaudou** [comment]
Hi,
I've just updated my GAMA Git and tried again the same use case.
I have a toolbar by default in the bottom of the fullscreen.

Now, when I just launched the simulation, the CMD+P and ESC keys do not work.
But after the click on any button of the bottom toolbar. (run, open of the side control, zoom in / out), the CMD+P / ESC work again.

Cheers

Benoit

**Comment by AlexisDrogoul** [comment]
Please see my comment above. The toolbar is now on by default unless you turn it off (in the model or in the preferences). Dont forget also that the fullscreen display, on two screens, is not the focus anymore. So it needs at least a click in the window to give it the keyboard focus (not simply move the mouse to it, although I could probably tweak it to follow the mouse, but let's solve one problem at a time).

**Comment by benoitgaudou** [comment]
Hi,

You 're right: with a click on the the fullscreen, the CMD+P / ESC work.
I just forget to give the focus to the fullscreen.



Cheers

Benoit

2018-07-27 18:00 GMT+02:00 Alexis Drogoul <notifications@github.com>:

**Comment by agrignard** [comment]
Hi everyone, 

Thanks for the hard work I am unfortunately in a very busy week so it's hard to follow this hot topic and all your comment

I just wanted to add something: this setting used to work on GAMA 1.7_XXX.YYY (let(s say around November 2017 when I created the old bug #2290 around the 16 of November 2017). Maybe it was working for a bad reason but it was working (actually the mediaLab demo is using a version prior to November 16 just because of that).

I will try next week to experiment a bit this settings and let you know

**Comment by agrignard** [comment]
Hi everyone!!! 

GOOOOOOOD NEWS!!!!! it's finally reworking as it was before. I am able to get the same settings as before. I am not sure to have understand everything in your conversation but with the last release GAMA1.8_Mac_64_07.28.18_89219fe I am now able to run it (the only change is to put the toolbar to false which is totally ok as soon as it's documented)

I guess we can close this issue? I will test it in real condition next week in the lab but from my settings at home computer+hdmi projector it sounds that this bug has been killed! good job!

**Comment by AlexisDrogoul** [comment]
Hi, thanks. One of the things not to forget is that, since "the last time", we have been more or less forced to migrate to oxygen, which has impacted a lot the display mechanisms at different levels. So lets close this, and open new issues on more focused issues if there are some in fullscreen displays !

**Comment by agrignard** [comment]
I reopen this one as I notice something that I can easily reproduce

(it might be linked to this issue https://github.com/gama-platform/gama/issues/2527 as I think it appears just after this enhancement) 

With this model 

`````
model multiscreen

global {
 init{
 	create people number:100;
 }
}

species people{
	aspect base{
		draw circle(1);
	}
}
experiment multiscreen type: gui{
	output {		
		
        display CityScope type:opengl{
        	species people aspect: base;
        }
		display CityScopeTable   type:opengl fullscreen:1 {
			species people aspect: base;
		}
	}
}

`````

The first time you open GAMA and the first time you run it you get this 
![img_3157](https://user-images.githubusercontent.com/3928502/44673167-b6e14400-a9f8-11e8-8930-4939d933874d.jpg)

If you relaunch or rerun the model then it's fine

**Comment by agrignard** [comment]
move to #2533
